### PR TITLE
*: show processlist

### PR DESCRIPTION
**Pull request by tiancaiamao**
Implement show processlist statement

```
mysql> show processlist;
+------+------+-----------+------+---------+------+-------+------------------+
| Id   | User | Host      | db   | Command | Time | State | Info             |
+------+------+-----------+------+---------+------+-------+------------------+
|    1 | root | 127.0.0.1 | test | Query   |    0 | 2     | show processlist |
+------+------+-----------+------+---------+------+-------+------------------+
1 row in set (0.00 sec)
```

This will be used to support session cancel feature https://github.com/pingcap/tidb/issues/2528.
`show processlist` get the session id, and `kill query id` cancel the session.

### CODE REVIEW

File: session.go [hunk 1]
```diff
@@ -985,3 +999,19 @@ func (s *session) InitTxnWithStartTS(startTS uint64) error {
 	}
 	return nil
 }
+
+func (s *session) ShowProcess() util.ProcessInfo {
```
Comment (coocood) [inline session.go]: We should avoid race here.

File: session.go [hunk 2]
```diff
@@ -477,6 +492,23 @@ func (s *session) ParseSQL(sql, charset, collation string) ([]ast.StmtNode, erro
...
 	s.prepareTxnCtx()
 	startTS := time.Now()
+
+	// Update processinfo, ShowProcess() will use it.
+	pi := util.ProcessInfo{
+		ID:      s.sessionVars.ConnectionID,
+		DB:      s.sessionVars.CurrentDB,
+		Command: "Query",
+		Time:    uint64(startTS.Unix()),
+		State:   fmt.Sprintf("%d", s.Status()),
```
Comment (coocood) [inline session.go]: We save processInfo on every query, we better use integer type here.

File: server/conn.go [hunk 3]
```diff
@@ -715,8 +716,10 @@ func (cc *clientConn) writeResultset(rs ResultSet, binary bool, more bool) error
 
 	columns, err := rs.Columns()
 	if err != nil {
+		fmt.Println("fuck...!!", err)
```
Comment (shenli) [inline server/conn.go]: Remove this.

File: infoschema/tables.go [hunk 4]
```diff
@@ -46,6 +46,7 @@ const (
 	tableSessionVar    = "SESSION_VARIABLES"
 	tablePlugins       = "PLUGINS"
 	tableConstraints   = "TABLE_CONSTRAINTS"
+	tableProcessList   = "PROCESSLIST"
```
Comment (shenli) [inline infoschema/tables.go]: Do you need to build the column list for processlist table?

Comment (tiancaiamao) [inline infoschema/tables.go]: Oh, I did it before, but give up later. I'll clean up here.

**Comment by coocood** [comment]
LGTM

**Comment by shenli** [comment]
LGTM
Please clean process info when query end and handle idle sessions.

**Comment by tiancaiamao** [comment]
updated.
PTAL @coocood  @shenli

File: session.go [hunk 5]
```diff
@@ -555,9 +559,24 @@ func (s *session) Execute(sql string) ([]ast.RecordSet, error) {
...
 	}
+	if len(rs) > 0 {
+		rs[len(rs)-1] = wrapRecordSet{rs[len(rs)-1], s}
+	} else if rs == nil {
+		s.setProcessInfo("")
+	}
 	return rs, nil
 }
 
+type wrapRecordSet struct {
```
Comment (shenli) [inline session.go]: Why we need this structure?

Comment (tiancaiamao) [inline session.go]: Close() seems to be a proper place to reset the processInfo, so I wrap a struct around the origin RecordSet, and overload it's Close() method to do it.

I also tried do it in CommitTxn or RollbackTxn, but it doesn't work well. The Info field of the results always get null, even the session that handle show processlist query.

Any suggestion here?

File: executor/adapter.go [hunk 6]
```diff
@@ -27,12 +27,17 @@ import (
 	"github.com/pingcap/tidb/plan"
 )
 
+type processinfoReset interface {
+	SetProcessInfo(string)
```
Comment (coocood) [inline executor/adapter.go]: `processinfoResetter` is more like an interface name.
And the method can be `ResetProcessInfo` without string argument.

**Comment by coocood** [comment]
LGTM

**Pull request merged**